sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala

```diff
@@ -2846,13 +2858,30 @@ private[lf] object SBuiltinFun {
     }
   }
 
-  private def importValue[Q](machine: Machine[Q], typ: Ast.Type, value: V)(
-      f: SValue => Control[Q]
+  /** Type-checks [createArg] against [dstTmplId] and converts it to an SValue. The [coid] and [srcTmplId] parameters
+    *  are used for error reporting only.
+    */
+  private def importCreateArg[Q](
+      machine: Machine[Q],
+      coidOpt: Option[V.ContractId],
+      srcTmplId: TypeConId,
+      dstTmplId: TypeConId,
+      createArg: V,
+  )(
+      k: SValue => Control[Q]
   ): Control[Q] = {
-    machine.importValue(typ, value) match {
-      case Right(value) => f(value)
-      case Left(error) => Control.Error(error)
-    }
+    new ValueTranslator(machine.compiledPackages.pkgInterface, forbidLocalContractIds = true)
+      .translateValue(Ast.TTyCon(dstTmplId), createArg)
+      .fold(
+        translationError =>
+          Control.Error(
+            IE.Upgrade(
+              IE.Upgrade
+                .TranslationFailed(coidOpt, srcTmplId, dstTmplId, createArg, translationError)
```

Comment:
It seems to me that many possibly unrelated errors are now caught by the upgrading umbrella (e.g. lookup errors, deeply nested values, unsuffixed c-ids etc.) so ideally we'd differentiate so users don't go looking in the wrong place. But I see that this might not be trivial so leaving this as wishful thinking